rust/sealevel/client/src/router.rs

```diff
@@ -0,0 +1,591 @@
+
+    /// Gets an instruction to set the owner.
+    fn set_owner_instruction(
+        &self,
+        client: &RpcClient,
+        program_id: &Pubkey,
+        new_owner: Option<Pubkey>,
+    ) -> Instruction;
+}
+
+pub(crate) trait ConnectionClient: Ownable {
+    /// Gets the interchain security module configured on-chain.
+    fn get_interchain_security_module(
+        &self,
+        client: &RpcClient,
+        program_id: &Pubkey,
+    ) -> Option<Pubkey>;
+
+    /// Gets an instruction to set the interchain security module.
+    fn set_interchain_security_module_instruction(
+        &self,
+        client: &RpcClient,
+        program_id: &Pubkey,
+        ism: Option<Pubkey>,
+    ) -> Instruction;
+}
+
+/// Idempotently deploys routers on multiple Sealevel chains and enrolls all routers (including
+/// foreign deployments) on each Sealevel chain.
+#[allow(clippy::too_many_arguments)]
+pub(crate) fn deploy_routers<
+    Config: for<'a> Deserialize<'a> + RouterConfigGetter + std::fmt::Debug + Clone,
+    Deployer: RouterDeployer<Config>,
+>(
+    ctx: &mut Context,
+    deployer: Deployer,
+    app_name: &str,
+    deploy_name: &str,
+    app_config_file_path: PathBuf,
+    chain_config_file_path: PathBuf,
+    environments_dir_path: PathBuf,
+    environment: &str,
+    built_so_dir_path: PathBuf,
+) {
+    // Load the app configs from the app config file.
+    let app_config_file = File::open(app_config_file_path).unwrap();
+    let app_configs: HashMap<String, Config> = serde_json::from_reader(app_config_file).unwrap();
+
+    // Load the chain configs from the chain config file.
+    let chain_config_file = File::open(chain_config_file_path).unwrap();
+    let chain_configs: HashMap<String, ChainMetadata> =
+        serde_json::from_reader(chain_config_file).unwrap();
+
+    let environments_dir = create_new_directory(&environments_dir_path, environment);
+
+    let artifacts_dir = create_new_directory(&environments_dir, app_name);
+    let deploy_dir = create_new_directory(&artifacts_dir, deploy_name);
+    let keys_dir = create_new_directory(&deploy_dir, "keys");
+
+    let existing_program_ids = read_router_program_ids(&deploy_dir);
+
+    // Builds a HashMap of all the foreign deployments from the app config.
+    // These domains with foreign deployments will not have any txs / deployments
+    // made directly to them, but the routers will be enrolled on the other chains.
+    let foreign_deployments = app_configs
+        .iter()
+        .map(|(chain_name, app_config)| (chain_name, app_config.router_config()))
+        .filter(|(_, router_config)| router_config.foreign_deployment.is_some())
+        .map(|(chain_name, router_config)| {
+            let chain_config = chain_configs.get(chain_name).unwrap();
+            (
+                chain_config.domain_id(),
+                hex_or_base58_to_h256(router_config.foreign_deployment.as_ref().unwrap()).unwrap(),
+            )
+        })
```

Comment:
nit: can use `filter_map` here and do `router_config.foreign_deployment.and_then(** closure passed to second "map" ** )`